SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduledJobPoller.java

```diff
@@ -28,110 +37,161 @@
+                    configuration.getWarnIfScheduledJobIsRunningPastNextRunPct());
+            }
+        }
     }
-  }
-
-  private Optional<Long> getExpectedRuntime(SingularityRequestWithState request, SingularityTaskId taskId) {
-    if (request.getRequest().getScheduledExpectedRuntimeMillis().isPresent()) {
-      return request.getRequest().getScheduledExpectedRuntimeMillis();
-    } else {
-      final Optional<SingularityDeployStatistics> deployStatistics = deployManager.getDeployStatistics(taskId.getRequestId(), taskId.getDeployId());
-
-      if (deployStatistics.isPresent() && deployStatistics.get().getAverageRuntimeMillis().isPresent()) {
-        return deployStatistics.get().getAverageRuntimeMillis();
-      }
-
-      final CronExpression cronExpression;
-
-      try {
-        cronExpression = new CronExpression(request.getRequest().getQuartzScheduleSafe());
-      } catch (ParseException e) {
-        LOG.warn("Unable to parse cron for {} ({})", taskId, request.getRequest().getQuartzScheduleSafe(), e);
-        exceptionNotifier.notify(e, ImmutableMap.of("taskId", taskId.toString()));
-        return Optional.absent();
-      }
-
-      final Date startDate = new Date(taskId.getStartedAt());
-      final Date nextRunAtDate = cronExpression.getNextValidTimeAfter(startDate);
-
-      if (nextRunAtDate == null) {
-        String msg = String.format("No next run date found for %s (%s)", taskId, request.getRequest().getQuartzScheduleSafe());
-        LOG.warn(msg);
-        exceptionNotifier.notify(msg, ImmutableMap.of("taskId", taskId.toString()));
-        return Optional.absent();
-      }
-
-      return Optional.of(nextRunAtDate.getTime() - taskId.getStartedAt());
+
+    private Optional<Long> getExpectedRuntime(SingularityRequestWithState request, SingularityTaskId taskId) {
+        if (request.getRequest().getScheduledExpectedRuntimeMillis().isPresent()) {
+            return request.getRequest().getScheduledExpectedRuntimeMillis();
+        } else {
+            final Optional<SingularityDeployStatistics> deployStatistics = deployManager.getDeployStatistics(taskId.getRequestId(), taskId.getDeployId());
+
+            if (deployStatistics.isPresent() && deployStatistics.get().getAverageRuntimeMillis().isPresent()) {
+                return deployStatistics.get().getAverageRuntimeMillis();
+            }
+            long nextRunAtTimestamp = 0;
+            final long now = System.currentTimeMillis();
+            try {
+                if (request.getRequest().getScheduleTypeSafe() == ScheduleType.RFC5545) {
+                    final String schedule = request.getRequest().getSchedule().get();
+                    final RecurrenceRule recurrenceRule = new RecurrenceRule(schedule);
+                    if (recurrenceRule.isInfinite()) {
+                        // set limit at 2100-01-01 00:00:00
+                        recurrenceRule.setUntil(new DateTime(2100, 1, 1, 0, 0, 0));
```

Comment:
Not really necessary, I put that just in case there was an infinite loop some how due to user input.

Got a better idea?